cmd/kubeadm/app/phases/upgrade/compute.go

```diff
@@ -0,0 +1,252 @@
+type Upgrade struct {
+	Description string
+	Before      StateBeforeUpgrade
+	After       StateAfterUpgrade
+}
+
+func (u *Upgrade) CanUpgradeKubelets() bool {
+	// If there are multiple different versions now, an upgrade is possible (even if only for a subset of the nodes)
+	if len(u.Before.KubeletVersions) > 1 {
+		return true
+	}
+	// Don't report something available for upgrade if we don't know the current state
+	if len(u.Before.KubeletVersions) == 0 {
+		return false
+	}
+
+	// if the same version number existed both before and after, we don't have to upgrade it
+	_, sameVersionFound := u.Before.KubeletVersions[u.After.KubeletVersion]
+	return !sameVersionFound
+}
+
+type StateBeforeUpgrade struct {
+	KubeVersion, DNSVersion, KubeadmVersion string
+	KubeletVersions                         map[string]int32
+}
+
+type StateAfterUpgrade struct {
+	KubeVersion, DNSVersion, KubeadmVersion, KubeletVersion string
+}
+
+// GetAvailableUpgrades fetches all versions from the specified versionGetter and computes which
+// kinds of upgrades can be performed
+func GetAvailableUpgrades(versionGetterImpl versionGetter, experimentalUpgradesAllowed, rcUpgradesAllowed bool) ([]Upgrade, error) {
+	fmt.Println("[upgrade] Fetching available versions to upgrade to:")
+
+	// Collect the upgrades kubeadm can do in this list
+	upgrades := []Upgrade{}
+
+	// Get the cluster version
+	clusterVersionStr, clusterVersion, err := versionGetterImpl.GetClusterVersion()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get current kubeadm CLI version
+	kubeadmVersionStr, kubeadmVersion, err := versionGetterImpl.GetKubeadmVersion()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get and output the current latest stable version
+	stableVersionStr, stableVersion, err := versionGetterImpl.GetLatestKubeVersion("stable", "stable version")
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the kubelet versions in the cluster
+	kubeletVersions, err := versionGetterImpl.GetKubeletVersions()
+	if err != nil {
+		return nil, err
+	}
+
+	// Construct a descriptor for the current state of the world
+	beforeState := StateBeforeUpgrade{
+		KubeVersion:     clusterVersionStr,
+		DNSVersion:      constants.GetKubeDNSVersion(clusterVersion),
+		KubeadmVersion:  kubeadmVersionStr,
+		KubeletVersions: kubeletVersions,
+	}
+
+	// Do a "dumb guess" that a new minor upgrade is available just because the latest stable version is higher than the cluster version
+	// This guess will be corrected once we know if there is a patch version available
+	canDoMinorUpgrade := clusterVersion.LessThan(stableVersion)
+
+	// A patch version doesn't exist if the cluster version is higher than the stable version
+	// In the case an user is trying to upgrade from, let's say, v1.8.0-beta.2 to v1.8.0-rc.1 (given we support such upgrades experimentally)
+	// a stable-1.8 branch doesn't exist yet. Hence this check.
+	if patchVersionBranchExists(clusterVersion, stableVersion) {
```

Comment:
if they want to upgrade from v1.8.0-beta.2 to v1.8.0-rc.1 _after_ v1.8 is released, I assume the current stable version would be `v1.8`, meaning that this conditional would pass, right?